queries/faust/highlights.scm

```diff
@@ -0,0 +1,216 @@
+; Identifiers
+(identifier) @variable
+
+(process) @variable.builtin
+
+(parameters
+  (identifier)) @variable.parameter
+
+; Literals
+(_
+  filename: (string)) @string.special.path
+
+(function_names) @string.special
+
+(documentation) @string.documentation @spell
+
+[
+  (string)
+  (fstring)
+] @string
+
+(real) @number.float
+
+[
+  (int)
+  (real)
+] @number
+
+; Types
+(global_metadata
+  key: (identifier)) @property
+
+(function_metadata
+  key: (identifier)) @property
+
+(_
+  type:
+    [
+      (int)
+      (float)
+      (any)
+    ]) @type.builtin
+
+[
+  (single_precision)
+  (double_precision)
+  (quad_precision)
+  (fixed_point_precision)
+] @attribute
+
+; Functions
+(function_definition
+  name: (identifier) @function)
+
+[
+  (lambda)
+  (prefix)
+  (prim1)
+  (prim2)
+  (prim3)
+  (prim4)
+  (prim5)
+  (function_call)
+] @function.call
+
+[
+  (exp)
+  (log)
+  (log10)
+  (sqrt)
+  (abs)
+  (floor)
+  (ceil)
+  (rint)
+  (round)
+  (acos)
+  (asin)
+  (atan)
+  (cos)
+  (sin)
+  (tan)
+  (atan2)
+  (int_cast)
+  (float_cast)
+  (pow_fun)
+  (min)
+  (max)
+  (fmod)
+  (remainder)
+  (prefix_prim)
+  (attach)
+  (enable)
+  (control)
+  (rdtable)
+  (rwtable)
+  (select2)
+  (select3)
+  (component)
+  (library)
+  (lowest)
+  (highest)
+  (assertbounds)
+  (par)
+  (seq)
+  (sum)
+  (prod)
+  (vslider_type)
+  (hslider_type)
+  (nentry_type)
+  (vbargraph_type)
+  (hbargraph_type)
+  (vgroup_type)
+  (hgroup_type)
+  (tgroup_type)
+  "button"
+  "checkbox"
+  "soundfile"
+  "inputs"
+  "outputs"
+  "route"
+] @function.builtin
```

Comment:
Do each of the builtin functions need a separate node type? This could be a limitation for writing queries. In many other languages built-in functions are managed as (a list of) literal strings. (Using `#any-of?`)

Reply:
Yeah I'm experimenting with being detailed enough to support using the AST for generating other downstream intermediates, e.g. with MLIR. I know it's not the intended use-case for tree-sitter, but the idea of only maintaining one parser/lexer is too appealing not to try :)

Reply:
I am sorry I don't get your point. What's blocking here from fixing the grammar and what does have it to do with 'one' lexers/parser?

Reply:
My goal is to support a consumer of the AST being able to, in theory, fully compile the AST into its corresponding Faust program. E.g. instead of having a Flex grammar/Bison parser that tokenizes and matches tokens to execute actions on those matches, I want to enable walking the tree-sitter generated AST instead and executing actions to compile the program. So the AST needs to somehow hold the information that a built-in function is e.g. a `cos` instead of a `sin`.

Reply:
I'm open to ideas for better ways of achieving this!

Reply:
Like if there was a way to walk into hidden nodes, that would be perfect

Reply:
tree-sitter generated AST would also allow visiting the token or the literal string of the identifiers or function name if needed. TS can visit unnamed nodes too. In my personal opinion, I am not fully convinced why they need to be "types" in the grammar.

Reply:
Interesting, I didn't know you could visit unnamed nodes. One thing that bugs me is that unnamed rules/fields can't be tested in the corpus since they don't show up. So you have untested outputs of the grammar that are only meant as... metadata for programmatically visiting the AST and not conceptually part of the AST itself? I see this is common practice (e.g. all the 'operator' named fields in the cpp grammar). Do folks usually maintain a separate test corpus to programmatically visit the expected unnamed nodes/fields?

What are the downsides of naming more rules than are strictly necessary? My understanding is it's mostly about performance?

Reply:
Performance, yes. State count and conflicts. You can run `./scripts/check-queries.lua faust` to get some timings for parsing the queries. See also https://github.com/tree-sitter/tree-sitter/wiki/Tips-and-Tricks-for-a-grammar-author

Reply:
This aspect has been the most challenging part of tree-sitter for me to make design decisions about, given my dual-use interests (incremental, real-time highlighting + first stage of compiler infrastructure). I've been thinking about it more with your valuable feedback in mind. Regardless of my issues with losing easy testability with unnamed/hidden nodes/fields, I will adapt my approach and
1. Rewrite my operators/terminal primitives to generally not be named nodes, in compliance with standard tree-sitter conventions.
2. Continue pursuing my hopes to use tree-sitter to generate a more detailed AST for compiling purposes, but do this in two passes - first, a "low res" pass using unnamed nodes conventionally, and then a programmatic transformation step walking through the tree, visiting unnamed nodes to decorate the low-res AST with more detail. I realize for my purposes, I'll need much more than I can expect to put into tree-sitter anyway (such as full literal values and variable names).

I'll make those grammar changes soon and post back - thanks for the feedback!

Reply:
@clason Do you have any suggestions for a case like this?
```
foo = f : + ~ g;
```
`~` is the "recursive" composition operator, and ':' is sequential composition. Before, this included the add operator on the lhs of the recursion:
```
  (definition
    variable: (identifier)
    value: (expression: (sequential
        left: (identifier)
        right: (recursive left: (add) right: (identifier))))
```
If I turn add into an unnamed '+' string, I lose it, which seems acceptable for binary infix operations, but here it seems wrong to take it out of the syntax tree:
```
  (definition
    variable: (identifier)
    value: (expression: (sequential
        left: (identifier)
        right: (recursive right: (identifier))))
```

Reply:
Honestly? You're asking the wrong person. We use tree-sitter for highlighting (and folding, and indenting), which requires low latency. A full syntax parser is very much a non-goal for us.

Reply:
Roger. I am working to give up on the full syntax parsing, as I explained in my comment, and I don't expect you to help me accomplish that goal, but I'm trying to find a way to do that and still keep it sensible. Do you consider this feedback of yours on having too many operator/primitive named rules a blocker for getting my grammar/highlighter used nvim-treesitter because of the performance implications? Or can I stick with my current grammar while I work on ways to strike a better balance?

Reply:
Well, the parser is all yours; these are feedbacks on the queries, which we'll have to maintain. I personally will not block anything.

As I wrote, you can check the performance with `./scripts/check-queries.lua`.

Reply:
@clason I updated the grammar and highlights to not used named rules for almost every built-in function.